libcrux-ml-dsa/src/simd/portable/arithmetic.rs

```diff
@@ -9,22 +9,79 @@ use crate::{
 pub(crate) const MONTGOMERY_SHIFT: u8 = 32;
 
 #[inline(always)]
+#[hax_lib::fstar::options("--z3rlimit 150")]
+#[hax_lib::requires(fstar!(r#"forall i. i < 8 ==> 
+    Spec.Utils.is_intb (pow2 31 - 1) (v (Seq.index ${lhs}.f_values i) + v (Seq.index ${rhs}.f_values i))"#))]
+#[hax_lib::ensures(|result| fstar!(r#"forall i. i < 8 ==>
+    (v (Seq.index ${lhs}_future.f_values i) == 
+     v (Seq.index ${lhs}.f_values i) + v (Seq.index ${rhs}.f_values i))"#))]
 pub fn add(lhs: &mut Coefficients, rhs: &Coefficients) {
+    let _lhs0 = lhs.clone();
```

Comment:
This may or may not be ok. But it will certainly increase the debug stack usage. For release builds these will get eliminated. But not in debug.

Reply:
The cloning is because the argument is passed by reference and not by value, but I assume it has the same cost. The loop invariant needs the original value to keep track on the not-yet-modified elements of the array.

Reply:
The solution would be to return a new array that has the modified values but AFAIK Rust allocates these arrays on stack too.

Reply:
Yep, we can't return values here. These have to be mutable arguments. We can probably tolerate a few of these clones. But probably not enough to do it in every function. So it's not an issue for this PR, but this style of proof. @karthikbhargavan lets look at this.

Reply:
This is now noted in https://github.com/cryspen/libcrux/issues/777
And I added a small style example for ghost state there, which we could implement within hax (it has to be handled in the engine in a specific way.)